torch/lib/THD/base/data_channels/Store.cpp

```diff
@@ -0,0 +1,187 @@
+#include "Store.hpp"
+#include "../ChannelUtils.hpp"
+
+#include <poll.h>
+#include <system_error>
+
+namespace thd {
+
+namespace {
+
+using store_type = std::unordered_map<std::string, std::vector<char>>;
+
+static void store_thread_daemon();
+static bool query(rank_type rank, const std::vector<int>& sockets);
+static bool checkAndUpdate(const store_type& store,
+                           std::vector<std::string>& keys);
+static void wake_up(int socket);
+
+enum class query_type : uint8_t {
+  set,
+  get,
+  wait,
+  finish,
+  stop_waiting
+};
+
+} // anonymous namespace
+
+Store::Store()
+    : _rank(load_rank_env()) {
+  // Master runs the store_thread
+  if (_rank == 0) {
+    _store_thread = std::thread(store_thread_daemon);
+    std::tie(_store_port, std::ignore) = load_master_env();
```

Comment:
This sounds more like a race condition than being ok. They're going to be initialized simultaneously.